src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,918 @@
+      implementFullBIP44: false,
+    });
+  }
+
+  async deserialize(opts: Partial<LedgerBridgeKeyringOptions> = {}) {
+    this.hdPath = opts.hdPath ?? hdPathString;
+    this.bridgeUrl = opts.bridgeUrl ?? BRIDGE_URL;
+    this.accounts = opts.accounts ?? [];
+    this.accountDetails = opts.accountDetails ?? {};
+    if (!opts.accountDetails) {
+      this.#migrateAccountDetails(opts);
+    }
+
+    this.implementFullBIP44 = opts.implementFullBIP44 ?? false;
+
+    // Remove accounts that don't have corresponding account details
+    this.accounts = this.accounts.filter((account) =>
+      Object.keys(this.accountDetails).includes(
+        ethUtil.toChecksumAddress(account),
+      ),
+    );
+
+    return Promise.resolve();
+  }
+
+  #migrateAccountDetails(opts: Partial<LedgerBridgeKeyringOptions>) {
+    if (this.#isLedgerLiveHdPath() && opts.accountIndexes) {
+      for (const account of Object.keys(opts.accountIndexes)) {
+        this.accountDetails[account] = {
+          bip44: true,
+          hdPath: this.#getPathForIndex(opts.accountIndexes[account] as number),
+        };
+      }
+    }
+
+    // try to migrate non-LedgerLive accounts too
+    if (!this.#isLedgerLiveHdPath()) {
+      this.accounts
+        .filter(
+          (account) =>
+            !Object.keys(this.accountDetails).includes(
+              ethUtil.toChecksumAddress(account),
+            ),
+        )
+        .forEach((account) => {
+          try {
+            this.accountDetails[ethUtil.toChecksumAddress(account)] = {
+              bip44: false,
+              hdPath: this.#pathFromAddress(account),
+            };
+          } catch (error) {
+            console.log(`failed to migrate account ${account}`);
+          }
+        });
+    }
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk?.publicKey);
+  }
+
+  isConnected() {
+    return this.isDeviceConnected;
+  }
+
+  setAccountToUnlock(index: number | string) {
+    this.unlockedAccount =
+      typeof index === 'number' ? index : parseInt(index, 10);
+  }
+
+  setHdPath(hdPath: string) {
+    // Reset HDKey if the path changes
+    if (this.hdPath !== hdPath) {
+      this.hdk = new HDKey();
+    }
+    this.hdPath = hdPath;
+  }
+
+  async unlock(hdPath?: string, updateHdk = true): Promise<string> {
+    if (this.isUnlocked() && !hdPath) {
+      return Promise.resolve('already unlocked');
+    }
+    const path = hdPath ? this.#toLedgerPath(hdPath) : this.hdPath;
+    return new Promise((resolve, reject) => {
+      this.#sendMessage(
+        {
+          action: IFrameMessageAction.LedgerUnlock,
+          params: {
+            hdPath: path,
+          },
+        },
+        ({ success, payload }) => {
+          if (success && isGetAddressMessageResponse(payload)) {
+            if (updateHdk) {
+              this.hdk.publicKey = Buffer.from(payload.publicKey, 'hex');
+              this.hdk.chainCode = Buffer.from(
+                payload.chainCode as string,
```

Comment:
Just noticed this — any way to avoid this typecast?

Reply:
Making that suggested type change would make sense to me. Then the assumption is at least documented in the type